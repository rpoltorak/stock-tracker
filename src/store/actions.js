export const ActionTypes = {
  ADD_COMPANY: "ADD_COMPANY",
  REMOVE_COMPANY: "REMOVE_COMPANY",
};
