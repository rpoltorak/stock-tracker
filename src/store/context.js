import React from "react";

export const Store = React.createContext({
  companies: {
    byId: {},
    ids: [],
  },
});
