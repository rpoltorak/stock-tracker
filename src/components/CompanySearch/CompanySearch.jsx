import React, { Fragment, useState, useContext, useEffect } from "react";
import { ActionTypes, Store } from "../../store";
import { useRequest } from "../../hooks";
import { search as searchCompany } from "../../services";
import { useDebounce } from "../../hooks";

export function CompanySearch() {
  const { dispatch } = useContext(Store);
  const [query, setQuery] = useState("");
  const [selectedItem, setSelectedItem] = useState(null);
  const [{ data, isLoading, isError }, setParams, setData] = useRequest(
    searchCompany(query),
    {
      bestMatches: [],
    },
  );
  const debouncedQuery = useDebounce(query, 500);

  useEffect(() => {
    if (debouncedQuery) {
      setParams(searchCompany(debouncedQuery));
    }
  }, [debouncedQuery]);

  return (
    <Fragment>
      <div>Company symbol</div>
      <form
        onSubmit={event => {
          dispatch({ type: ActionTypes.ADD_COMPANY, payload: selectedItem });
          setQuery("");
          setData({
            bestMatches: [],
          });
          event.preventDefault();
        }}
      >
        <input
          type="text"
          value={query}
          onChange={event => setQuery(event.target.value)}
          data-testid="search"
        />
        <button type="submit">Track</button>
      </form>
      {isError && <div>Error occured</div>}
      {isLoading ? (
        <div>Loading ...</div>
      ) : (
        <ul>
          {data.bestMatches &&
            data.bestMatches.map(({ name, symbol }) => (
              <li key={symbol}>
                <a
                  href="#"
                  onClick={() => {
                    setQuery(symbol);
                    setSelectedItem({ name, symbol });
                  }}
                >
                  {symbol}: {name}
                </a>
              </li>
            ))}
        </ul>
      )}
      <div>
        Provide the stock exchange symbol of a company you want to track
      </div>
    </Fragment>
  );
}
