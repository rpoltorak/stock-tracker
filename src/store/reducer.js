import omitBy from "lodash.omitby";
import { ActionTypes } from "./actions";

const initialState = {
  companies: {
    byId: {},
    ids: [],
  },
};

export function reducer(state = initialState, action) {
  switch (action.type) {
    case ActionTypes.ADD_COMPANY:
      return addCompany(state, action.payload);
    case ActionTypes.REMOVE_COMPANY:
      return removeCompany(state, action.payload);
    default:
      return state;
  }
}

function addCompany(state, company) {
  const { symbol } = company;

  if (!symbol || state.companies.ids.includes(symbol)) {
    return state;
  }

  return {
    ...state,
    companies: {
      byId: {
        ...state.companies.byId,
        [symbol]: company,
      },
      ids: [...state.companies.ids, symbol],
    },
  };
}

function removeCompany(state, symbol) {
  return {
    companies: {
      byId: omitBy(state.companies.byId, item => item.symbol === symbol),
      ids: state.companies.ids.filter(item => item !== symbol),
    },
  };
}
