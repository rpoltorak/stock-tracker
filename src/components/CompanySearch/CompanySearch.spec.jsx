import React from "react";
import axios from "axios";
import { render, fireEvent, wait, cleanup } from "@testing-library/react";
import { Store } from "../../store";
import { CompanySearch } from "./CompanySearch";

const setup = () => {
  const utils = render(<CompanySearch />);
  const input = utils.getByTestId("search");
  return {
    input,
    ...utils,
  };
};

const exampleData = [
  {
    "1. symbol": "GOOG",
    "2. name": "Alphabet Inc.",
    "3. type": "Equity",
    "4. region": "United States",
    "5. marketOpen": "09:30",
    "6. marketClose": "16:00",
    "7. timezone": "UTC-04",
    "8. currency": "USD",
    "9. matchScore": "1.0000",
  },
  {
    "1. symbol": "GOOGL",
    "2. name": "Alphabet Inc.",
    "3. type": "Equity",
    "4. region": "United States",
    "5. marketOpen": "09:30",
    "6. marketClose": "16:00",
    "7. timezone": "UTC-04",
    "8. currency": "USD",
    "9. matchScore": "0.8889",
  },
];

jest.mock("axios");

afterEach(cleanup);

describe("CompanySearch", () => {
  test("renders properly on idle state", () => {
    const { getByText, input } = setup();

    expect(getByText("Company symbol")).toBeTruthy();
    expect(input).toBeTruthy();
  });

  test("makes an search api call after input change", async () => {
    axios.mockResolvedValue({ data: { bestMatches: exampleData } });

    const { input } = setup();

    fireEvent.change(input, { target: { value: "GOOG" } });

    await wait(() => {
      expect(axios).toHaveBeenCalled();
    });
  });

  test("renders the result after input change", async () => {
    const { input, container } = setup();

    fireEvent.change(input, { target: { value: "GOOG" } });

    await wait(() => {
      expect(container.querySelector(".list-group-item")).toBeTruthy();
      expect(container.querySelectorAll(".list-group-item").length).toEqual(2);
    });
  });

  test("dispatches an action after clicking track button", async () => {
    axios.mockResolvedValue({ data: { bestMatches: exampleData } });

    const state = {
      companies: {
        byId: {},
        ids: [],
      },
    };
    const dispatch = jest.fn();

    const { getByTestId, getByText, container } = render(
      <Store.Provider value={{ state, dispatch }}>
        <CompanySearch />
      </Store.Provider>,
    );

    fireEvent.change(getByTestId("search"), { target: { value: "GOOG" } });

    await wait(() => {
      fireEvent.click(container.querySelectorAll(".list-group-item")[0]);
      fireEvent.click(getByText("Track"));
    });

    expect(dispatch).toHaveBeenCalled();
  });
});
