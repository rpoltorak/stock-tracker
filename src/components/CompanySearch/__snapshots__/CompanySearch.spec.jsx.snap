// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompanySearch matches the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="mt-4"
      >
        <form
          class=""
        >
          <div
            class="form-group"
          >
            <div
              class="row"
            >
              <div
                class="col-sm-5"
              >
                <label
                  class="form-label"
                >
                  Company symbol
                </label>
                <input
                  class="form-control"
                  data-testid="search"
                  placeholder="Company symbol"
                  type="text"
                  value=""
                />
                <small
                  class="text-muted form-text"
                >
                  Provide the stock exchange symbol of a company you want to track
                </small>
              </div>
            </div>
          </div>
          <button
            class="btn btn-primary"
            disabled=""
            type="submit"
          >
            Track
          </button>
        </form>
        <div
          class="mt-4 list-group"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="mt-4"
    >
      <form
        class=""
      >
        <div
          class="form-group"
        >
          <div
            class="row"
          >
            <div
              class="col-sm-5"
            >
              <label
                class="form-label"
              >
                Company symbol
              </label>
              <input
                class="form-control"
                data-testid="search"
                placeholder="Company symbol"
                type="text"
                value=""
              />
              <small
                class="text-muted form-text"
              >
                Provide the stock exchange symbol of a company you want to track
              </small>
            </div>
          </div>
        </div>
        <button
          class="btn btn-primary"
          disabled=""
          type="submit"
        >
          Track
        </button>
      </form>
      <div
        class="mt-4 list-group"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "input": <input
    class="form-control"
    data-testid="search"
    placeholder="Company symbol"
    type="text"
    value=""
  />,
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
